components/client/src/main/java/com/hotels/styx/client/connectionpool/ConnectionUsageTracker.java

```diff
@@ -0,0 +1,64 @@
+/**
+ * Copyright (C) 2013-2018 Expedia Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hotels.styx.client.connectionpool;
+
+import com.hotels.styx.api.client.Connection;
+
+/**
+ * Provides methods used for decorating and tracking connections for additional conditions for expiration.
+ */
+interface ConnectionUsageTracker {
+
+    /**
+     * Used for decoration of a connection to provide expiration tracking logic.
+     * @param connection connection to decorate
+     * @return decorated connection
+     */
+    Connection decorate(Connection connection);
+
+    /**
+     * Provides logic for checking if connection is usable.
+     * @param connection connection under test.
+     * @return value whether the connection has expired.
+     */
+    boolean shouldTerminate(Connection connection);
```

Comment:
The `shouldTerminate` feels out of place. Like a leaking abstraction. So if we have a generic `ConnectionUsageTracker` interface behind which it is possible to implement arbitrary connection tracking logic, then why does it need `shouldTerminate` method?

I guess I am trying to say that `shouldTerminate` is very specific to a particular implementation of this interface. Therefore it shouldn't have place in an interface like this. Further evidence for this can be found in `ConnectionExpirationTracker` class where we need to resort to `instanceof` checks:

```
    @Override
    public boolean shouldTerminate(Connection connection) {
        return connection instanceof ExpirableConnection && ((ExpirableConnection) connection).isExpired();
    }
```

Perhaps better way to approach this problem is to perform the check in the `isConnected()` method of the adapting class (`TrackedConnectionAdapter`). Perform the check there and close the underlying channel if necessary. Then we could get away with this method and the associated instance checks.